src/local-mode.ts

```diff
@@ -1,12 +1,59 @@
 import * as Bluebird from 'bluebird';
 import * as Docker from 'dockerode';
 import * as _ from 'lodash';
-
 import Config from './config';
 import Database from './db';
+import log from './lib/supervisor-console';
 import { Logger } from './logger';
 
-import log from './lib/supervisor-console';
+// EngineSnapshot represents a list of containers, images, volumens, and networks present on the engine.
+// A snapshot is taken before entering local mode in order to perform cleanup when we exit local mode.
+export class EngineSnapshot {
+	constructor(
+		public readonly containers: string[],
+		public readonly images: string[],
+		public readonly volumes: string[],
+		public readonly networks: string[],
+	) {}
+
+	public static fromJSON(json: string): EngineSnapshot {
+		const obj = JSON.parse(json);
+		return new EngineSnapshot(
+			obj.containers,
+			obj.images,
+			obj.volumes,
+			obj.networks,
+		);
+	}
+
+	// Builds a new snapshot object that contains entities present in another snapshot,
+	// but not present in this one.
+	public diff(another: EngineSnapshot): EngineSnapshot {
+		return new EngineSnapshot(
+			_.difference(this.containers, another.containers),
+			_.difference(this.images, another.images),
+			_.difference(this.volumes, another.volumes),
+			_.difference(this.networks, another.networks),
+		);
+	}
+
+	public toString(): string {
+		return (
+			`${this.containers.length} containers, ` +
+			`${this.images.length} images, ` +
+			`${this.volumes.length} volumes, ` +
+			`${this.networks.length} networks`
+		);
+	}
+}
+
+// Record in a database that stores EngineSnapshot.
+export class EngineSnapshotRecord {
```

Comment:
I like this :+1: We should start to define all our models in this way for the typescript conversion.